### DISCUSSION TITLE
Which configuration for Daikin Rotex HPSU Compact 508 H/C BIV and how to implement cooling function relay

### DISCUSSION THREAD

**Discussion by DanieleNeo** [raomin/ESPAltherma#467] [General]
Hi,

very nice project your project ESPAltherma. I'm setting my own and I have a couple of question for which I would need your feedback. I have a Daikin Rotex HPSU Compact 508 H/C BIV and I found ([https://github.com/raomin/ESPAltherma/issues/45) ](https://github.com/raomin/ESPAltherma/issues/45))that I can use Altherma using the ALTHERMA(HPSU6_ULTRA).h include (the BIV version is the same of 508 with support for external heat sources), the problem is that I downloaded last sources and I'm unable to find this include, I have to use a different one, and if so which one?

Then, in the Getting started section there is "ESPAltherma can control a relay on MQTT that can simulate an external On Off thermostat. Doing so allows to remotely turn on/off the heating function of your heat pump. A second relay can be used to trigger the cooling function".  I would like to use the second relay in order to drive the cooling function, but how can I use it?

Thanks in advance,
Daniele.

**Comment by philipgo** [comment]
I could not get the second relay to work, so I used a second ESP32 device, see here: https://github.com/raomin/ESPAltherma/discussions/460

**Reply by DanieleNeo** [reply]
Hi @philipgo,

but you tried to modify the source and failed?  using a second ESP32 device could be a solution, anyway considering that most of heat pump have cooling function it would be desirable if such support was made available directly by ESPaltherma. As soon as I have a moment I'll try to take a look at it, even though C/C++ is not within my programming skills, I think the authors are able to implement it with little effort.

@raomin

**Reply by philipgo** [reply]
I tried and it did not compile immediately. I then did not spend much time troubleshooting because I did not even know how to connect two relays to my M5 StickC. 

The M5 dual relay modules @HarmEllis used do not have a "normally closed" position which was important to me. If my (ESPHome) heating relay fails, my Rotex will revert to "heating" until I fix the issue.

**Reply by DanieleNeo** [reply]
@philipgo,

of course yours is a solution, and if you like it that's fine. I prefer a minimal approach, which is why I started modifying the source to support a second relay, I share the patch with the community

[cooling.patch](https://github.com/user-attachments/files/17638911/cooling.patch)

This is the result
 
![image](https://github.com/user-attachments/assets/200a5e24-0991-4c98-b5db-af516fb552ff)

Also, always to follow a minimalist approach, to have everything more contained and limit energy consumption (to not have problems powering the board from the heat pump) instead of the relays I'm using two 4N35 optocouplers, their job (close the circuit) is done, I just have to try it on the system. As soon as I can I'll post some photos.

If @raomin wants this patch can be included in the source.

D.

**Reply by philipgo** [reply]
Looks very nice! I would also prefer a minimalist approach but am unable to do this kind of coding :) 

Two questions:
- does your code prevent both switches from being on at the same time?
- Can you describe the ESP32 hardware and wiring?

Thank you!

**Reply by DanieleNeo** [reply]
Hi @philipgo,

I think the real orchestration should be done in Home Assistant, but I still inserted a safety check to prevent a relay from being activated if there is already one active, I attach the patched version. I also corrected some texts as well as changed the default pins (2 and 4, respectively for heater and cooler), because I noticed that during the update pin 0 assumes a high value, and therefore the relay is activated.

As a board I use an ESP32 that I already had, I don't care about having a display. I think I'll publish diagrams and some photos in the next few days, I first have to make sure that the heat pump "digests" the 4N35 (I tried with a normal boiler, and with that it worked).

D.

[cooling2.patch](https://github.com/user-attachments/files/17822813/cooling2.patch)

**Reply by DanieleNeo** [reply]
Hi @philipgo,

as promised this is the wiring diagram

![espaltherma_2ndrelay_mod](https://github.com/user-attachments/assets/5b99f918-2664-44d1-ae2c-242c5d863350)

and some photos of realizzation and implementation

![IMG_20241106_212113](https://github.com/user-attachments/assets/c2234000-5c68-4b62-91af-ceb415d639a1)
![IMG_20241106_212106](https://github.com/user-attachments/assets/b44aa09e-5247-41cd-8a34-67b73190c244)
![IMG_20241106_212226](https://github.com/user-attachments/assets/d30c3fbc-2746-4d70-b6d1-59e39510d7e2)
![IMG_20241106_212236](https://github.com/user-attachments/assets/e7458009-1f0f-432d-b898-5766b78831a8)
![IMG_20241106_212509](https://github.com/user-attachments/assets/98ef1768-894f-437f-a8ed-8a2b4cc5fb06)
![IMG_20241106_212700](https://github.com/user-attachments/assets/7d411857-3691-4d49-a1c8-48b6141267b5)
![IMG_20241106_210626](https://github.com/user-attachments/assets/c28eddb2-b89e-4470-9d6c-042d2bf0adbc)

As you can see I assembled a board for convenience from which I can replace the ESP32. I still have to check the operation of the 4N35 optocoupler on the heat pump, as mentioned I did a test with positive results on a boiler, but for the heat pump I have to wait for it to be started in the next few days (it is an old hardware purchased some time ago and never put into operation).

The connection interface that I represented is from my heat pump (Daikin Rotex HPSU Compact 508 H/C BIV), but what is published on the espaltherma website also applies. At the moment as a configuration (includes) I am using the default, but does anyone recommend a more suitable one for my hardware? :)

D.

**Reply by DanieleNeo** [reply]
Obviously you can use signal from GPIO2 and 4 (or whatever you want) for driving a realay ;)

**Reply by DanieleNeo** [reply]
@philipgo, I created a pull request (https://github.com/raomin/ESPAltherma/pull/477) hoping this patch could be merged in main branch.
